src/budget_program.rs

```diff
@@ -230,32 +230,22 @@ impl BudgetState {
         deserialize(&input[8..8 + len as usize])
     }
 
-    fn save_error_to_budget_state(e: BudgetError, accounts: &mut [Account]) -> () {
-        if let Ok(mut state) = BudgetState::deserialize(&accounts[1].userdata) {
-            trace!("saved error {:?}", e);
-            state.last_error = Some(e);
-            state.serialize(&mut accounts[1].userdata).unwrap();
-        } else {
-            trace!("error in uninitialized contract {:?}", e,);
-        }
-    }
-
     /// Budget DSL contract interface
     /// * tx - the transaction
     /// * accounts[0] - The source of the tokens
     /// * accounts[1] - The contract context.  Once the contract has been completed, the tokens can
     /// be spent from this account .
-    pub fn process_transaction(tx: &Transaction, accounts: &mut [Account]) -> () {
+    pub fn process_transaction(
+        tx: &Transaction,
+        accounts: &mut [Account],
+    ) -> Result<(), BudgetError> {
         if let Ok(instruction) = deserialize(&tx.userdata) {
             trace!("process_transaction: {:?}", instruction);
-            let _ = Self::apply_debits_to_budget_state(tx, accounts, &instruction)
+            Self::apply_debits_to_budget_state(tx, accounts, &instruction)
                 .and_then(|_| Self::apply_credits_to_budget_state(tx, accounts, &instruction))
-                .map_err(|e| {
-                    trace!("saving error {:?}", e);
-                    Self::save_error_to_budget_state(e, accounts);
-                });
         } else {
             info!("Invalid transaction userdata: {:?}", tx.userdata);
+            Err(BudgetError::UserdataDeserializeFailure)
```

Comment:
This case is probably unreachable. If so, consider changing it to `unreachable!`.